### PR TITLE
.es_temp_file remains after system crash, causing it not to start again

### LINKED ISSUE #20992: [Bug Fix] .es_temp_file remains after system crash
When system starts, it creates a temporary file named .es_temp_file to ensure the data directories are writable.

If system crashes after creating the .es_temp_file but before deleting this file, next time the system will not be able to start because the `Files.createFile(resolve)` will throw an exception if the file already exists.

So I first check if the temp file exists, if yes, just delete it; if not exist, then create and delete it.

Please refer to the pull request #20991 for more details.

### PR DESCRIPTION
**Pull request by liweinan0423**
Closes #20992 

Move the delete to `finally` block to ensure the file will be deleted when there is a left over temp file from a crash.

### CODE REVIEW

**Comment by jasontedor** [comment]
test this please

**Comment by liweinan0423** [comment]
The Jenkins build seems contains many other commits?

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -1004,9 +1004,10 @@ private static void tryWriteTempFile(Path path) throws IOException {
             Path resolve = path.resolve(".es_temp_file");
             try {
                 Files.createFile(resolve);
-                Files.deleteIfExists(resolve);
             } catch (IOException ex) {
                 throw new IOException("failed to write in data directory [" + path + "] write permission is required", ex);
+            } finally {
+                Files.deleteIfExists(resolve);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: can we make sure not to delete a file that wasn't created and hide the original exception?

we typically do this:
```
bool success = false;
try {
  do something
  success = true;
} finally {
  if (success) {
    clean up
  }
}
```

**Review by bleskes — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java (hunk 1 above)
Comment (liweinan0423) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: Thanks for you advice. I have pushed a new commit.

**Comment by bleskes** [comment]
test this please

**Review by bleskes — APPROVED** [review]
LGTM. I'll merge it in. Thank you!!

**Comment by liweinan0423** [comment]
The tests are successful on CI - https://elasticsearch-ci.elastic.co/job/elastic-elasticsearch-pull-request/582/

**Comment by s1monw** [comment]
@bleskes this change is doing exactly the same thing it was doing before. It only deletes the file if the creation was successful. I think we should revert it and make it try to delete the file even if we run into an exception this can possibly prevent the issue but if the system crashes in between there is nothing we can do about it. Either way this change is only confusing.

**Comment by bleskes** [comment]
@s1monw bleh. I have no idea what I was thinking. What I was going for is not hide the original exception from the file creation. I'll fix.

**Comment by liweinan0423** [comment]
I suggest if we detect the file already exists, try to delete it first then create it again, then delete it in the finally block.

**Pull request merged**